libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -897,6 +899,71 @@ std::string YulUtilFunctions::resizeDynamicArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::resizeDynamicByteArrayFunction(ArrayType const& _type)
+{
+	string functionName = "resize_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(array, newLen) {
+				if gt(newLen, <maxArrayLength>) {
+					<panic>()
+				}
+
+				let oldLen := <fetchLength>(array)
+
+				if gt(newLen, oldLen) {
+					sstore(array, or(shl(newLen, 1), iszero(lt(newLen, 32))))
+				}
+
+				// Size was reduced, clear end of array
+				if lt(newLen, oldLen) {
+					let offset := mod(newLen, 32)
+					let oldSlotCount := add(iszero(oldLen), div(add(oldLen, 31), 32))
+					let newSlotCount := add(iszero(newLen), div(add(newLen, 31), 32))
+
+					switch eq(newSlotCount, 1)
+					case  0 {
+						let arrayDataStart := <dataPosition>(array)
+						let deleteStart := add(arrayDataStart, newSlotCount)
+						let deleteEnd := add(arrayDataStart, oldSlotCount)
+
+						// we have  to partially clear last slot that is still used, so decreasing start by one
+						if gt(offset, 0) { <partialClearStorageSlot>(sub(deleteStart, 1), offset) }
+
+						<clearStorageRange>(deleteStart, deleteEnd)
+
+						sstore(array, or(shl(1, newLen), 1))
+					}
+					default {
+						let arrayDataStart := array
+						if gt(oldSlotCount, 1) {
+							arrayDataStart := <dataPosition>(array)
+							// clear whole old array, as we are transforming to short bytes array
+							<clearStorageRange>(arrayDataStart, add(arrayDataStart, oldSlotCount))
+						}
+						// we need to copy elements from old array to new
+						let data := sload(arrayDataStart)
+						// we want to copy only elements that are part of the array after resizing
+						let mask := <shr>(mul(8, sub(32, offset)), <ones>)
+						data := and(data, mask)
+						sstore(array, or(data, <shl>(1, newLen)))
```

Comment:
We already have some byte array functions, can't we extract something here into a common helper function?

Reply:
Yes, it should be common helper. I think I use the same code in ByteArrayPopFunction
Will extract it now

Reply:
I moved to two functions, which are maybe to general to use in bytearray push and pop functions
Will maybe think one more time how to extract shared part.

Reply:
It might even make sense to extract very simple things like "set the length of a short byte array while not touching the data" and "set the length of a short byte array and providing the data"

Reply:
Will extract these smaller functions